### ISSUE TITLE
Padding on heterogenous intersection

### ISSUE THREAD

**Issue by thoithoi58** [dxing-cs/TinyLight#1]
Hi, 

I've just found out you repo and it worked flawlessly. However I have a question:

The original MPLight algo was tested on Manhattan map, which have some heterogenous intersection. The author said for those intersection need to be padded. But I don't know how to edit your code to do so. Can you help me?

Many thanks

**Comment by dxing-cs** [comment]
Hi, 

Thanks for your interest!

The easiest approach I can come up with is to trim the roadnet.json file to align all intersections. You can also achieve it by editing the forward function [here](https://github.com/dxing-cs/TinyLight/blob/6b5dee86ac02e3916242179901ed64acca0e85e8/TinyLight/agent/drl_based/frap.py#L73) to pad the observation with fewer items before it proceeds. 

> Hi,
> 
> I've just found out you repo and it worked flawlessly. However I have a question:
> 
> The original MPLight algo was tested on Manhattan map, which have some heterogenous intersection. The author said for those intersection need to be padded. But I don't know how to edit your code to do so. Can you help me?
> 
> Many thanks

**Comment by thoithoi58** [comment]
Hi,

I appriciate you response, however in the original paper. They said :  Each agent observes part of the system state as its own observation. For a standard intersection with 12 traffic movements, its observation includes the current phase p and the pressure of the 12 traffic movements. **Note that for the intersection with fewer than 12 movements, the vector is zero-padded**. And also changing the map or pad the observation with fewer items will no longer match the solution of the orginal paper. But I can't come up with where to pad the observation before feed it to the network. Can you suggest any idea?

Many thanks.




> Hi,
> 
> Thanks for your interest!
> 
> The easiest approach I can come up with is to trim the roadnet.json file to align all intersections. You can also achieve it by editing the forward function [here](https://github.com/dxing-cs/TinyLight/blob/6b5dee86ac02e3916242179901ed64acca0e85e8/TinyLight/agent/drl_based/frap.py#L73) to pad the observation with fewer items before it proceeds.
> 
> > Hi,
> > I've just found out you repo and it worked flawlessly. However I have a question:
> > The original MPLight algo was tested on Manhattan map, which have some heterogenous intersection. The author said for those intersection need to be padded. But I don't know how to edit your code to do so. Can you help me?
> > Many thanks

**Comment by dxing-cs** [comment]
Hi, 

You can always pad the observation in this way: 

```python 
def forward(self, obs): 
    obs[1] = torch.hstack([obs[1], torch.zeros(1, LENGTH_AFTER_PAD - obs[1].shape[1])])
    # the following is the same 
```

Here `obs[1]` corresponds to the movement features. In my re-implementation for the MPLight, the observation is composed of `lane_link`. Therefore, each movement corresponds to 3 `lane_link` in the setting above. Does this address your problem? I'm happy to explain if you need more help. 

> Hi,
> 
> I appriciate you response, however in the original paper. They said : Each agent observes part of the system state as its own observation. For a standard intersection with 12 traffic movements, its observation includes the current phase p and the pressure of the 12 traffic movements. **Note that for the intersection with fewer than 12 movements, the vector is zero-padded**. And also changing the map or pad the observation with fewer items will no longer match the solution of the orginal paper. But I can't come up with where to pad the observation before feed it to the network. Can you suggest any idea?
> 
> Many thanks.
> 
> > Hi,
> > Thanks for your interest!
> > The easiest approach I can come up with is to trim the roadnet.json file to align all intersections. You can also achieve it by editing the forward function [here](https://github.com/dxing-cs/TinyLight/blob/6b5dee86ac02e3916242179901ed64acca0e85e8/TinyLight/agent/drl_based/frap.py#L73) to pad the observation with fewer items before it proceeds.
> > > Hi,
> > > I've just found out you repo and it worked flawlessly. However I have a question:
> > > The original MPLight algo was tested on Manhattan map, which have some heterogenous intersection. The author said for those intersection need to be padded. But I don't know how to edit your code to do so. Can you help me?
> > > Many thanks
